### ISSUE TITLE
[semi-random/use-case] hierarchical config and promotion versions

### ISSUE THREAD

**Issue by aegershman** [pivotal-cf/om#362]
This isn't an issue; I was inspired by https://github.com/pivotal-cf/om/issues/360#issuecomment-490781915

I want to comment on @vchrisb 's use-case in [#360] without hijacking their issue. We do "hierarchical configuration monorepos"(?) && I want to share thoughts on how we promote config/versions from sandbox to dev to prod for all our IaaSes/regions/envs/foundations.

## current usage

Similar to what vchrisb described: we glob the `patch` version of pivnet product, download it to S3, then automation takes the _exact_ version of the latest that we just downloaded && opens a PR to a github repo, which starts this upgrade chain from env to env.

Here's a very stripped-down version of our `upgrade-tiles-config` configuration monorepository that holds information on the `product-version` and `stemcell-version` which should be applied for each foundation. I'm showing a limited set of configuration && limited subset of the `upgrade-tiles-config/` monorepo here:

```
pas
├── aws
│   └── us-west-2
│       ├── dev
│       │   ├── cf
│       │   │   ├── product-version
│       │   │   ├── stemcell-version
│       │   │   └── vars.yml
│       │   └── config
│       │       ├── auth.yml
│       │       ├── cf-secrets.yml
│       │       ├── director-secrets.yml
│       │       ├── env.yml
│       │       └── opsman.yml
│       ├── prod
│       │   ├── cf
│       │   │   ├── product-version
│       │   │   ├── stemcell-version
│       │   │   └── vars.yml
│       │   └── config
│       │       ├── auth.yml
│       │       ├── cf-secrets.yml
│       │       ├── director-secrets.yml
│       │       ├── env.yml
│       │       └── opsman.yml
│       └── sandbox
│           ├── cf
│           │   ├── product-version
│           │   ├── stemcell-version
│           │   └── vars.yml
│           └── config
│               ├── auth.yml
│               ├── cf-secrets.yml
│               ├── director-secrets.yml
│               ├── env.yml
│               └── opsman.yml
├── azure
│   └── centralus
│       ├── dev
│       │   ├── cf
│       │   │   ├── product-version
│       │   │   ├── stemcell-version
│       │   │   └── vars.yml
│       │   └── config
│       │       ├── auth.yml
│       │       ├── cf-secrets.yml
│       │       ├── director-secrets.yml
│       │       ├── env.yml
│       │       └── opsman.yml
│       └── sandbox
│           ├── cf
│           │   ├── product-version
│           │   ├── stemcell-version
│           │   └── vars.yml
│           └── config
│               ├── auth.yml
│               ├── cf-secrets.yml
│               ├── director-secrets.yml
│               ├── env.yml
│               └── opsman.yml
└── vsphere
    └── kc8
        └── dev
            ├── cf
            │   ├── product-version
            │   ├── stemcell-version
            │   └── vars.yml
            └── config
                ├── auth.yml
                ├── cf-secrets.yml
                ├── director-secrets.yml
                ├── env.yml
                └── opsman.yml
```

## workflow

- there's a separate pipeline which listens to pivnet for changes on product or stemcell globs (e.g. `cf-2.5.*` or `xenial-315.0.*`, etc.)
- once it saves it to s3, it opens a PR to the `upgrade-tiles-config` repo to update the product's version or stemcell (e.g., `cf/product-version` would be `2.5.3-build.17` and `cf/stemcell-version` would be `250.38.0`) within each IaaS's respective `sandbox` (or however you'd like to configure it)
- We review & merge in the automated PR, which kicks off a pipeline that performs a selective-deploy to the foundation (an entirely different discussion topic).
- Once that selective `upload/stage/config/apply` job completes, the pipeline **takes the product/stemcell versions that successfully passed `sandbox` and creates a PR to the next foundation**
- This is done because aws/sandbox concourse pipelines are configured with params that controls the PR's "target"; they look like:
```yml
# upgrade-tiles-config/pipelines/config/pas/aws/us-west-2/sandbox.yml
next_iaas_type_in_chain: aws
next_iaas_region_in_chain: us-west-2
next_foundation_in_chain: dev
```
- The process repeats for aws/dev: we merge the PR, automation does it's thing, then, because `dev` has a `next_foundation_in_chain` of `prod`, process repeats for prod.
- After `prod` is merged/applied, since it has a `next_foundation_in_chain: prod`, which is _itself_, that signals the PR process to stop.

## value: upgrades to the latest and greatest

We repeat this process for _all our environments_ across _all IaaSes_ every week, and have changes in `prod` by the end of the week-- and we successfully validated every change in both a `sandbox` and `dev` environment. When product-patch or stemcell upgrades hits `pivnet` on Monday, it's in prod by Friday as part of our weekly upgrade/repave cadence. Assuming all goes right in each foundation, using this automation pattern, we take upgrades at the speed of `git merge`.

## value: github labels

This process creates a lot of github PRs, but it's actually pretty great, because we add `labels` to each PR (`iaas: vsphere`, `upgrade: stemcell` if stemcell-version changes, `product-name: cf`, etc.). We can then use those labels as basic value-stream mapping ("how many stemcell upgrades have you taken?" just search by labels)

## some sort of actual pattern, maybe?

I'm sure there are all sorts of different ways to do this, but I hope a standardized pattern will eventually emerge where you define a configuration monorepo that acts like a state machine... sort of? 

You define "promotion rules" to transition from one state to other states (e.g., take `<some-file>` in `aws/us-west-2/sandbox/**` and promote it to `aws/us-west-2/dev/**` and `aws/us-east-1/dev**`. Then take `aws/us-west-2/dev/**` and promote it to `aws/us-west-2/prod/**`). 

You can configure "transitions" from state-to-state, so it's an extensible workflow. For example if `cf-isoseg` is only in my `aws/sandbox` and `aws/prod` foundations (NOT in `aws/dev`), then for `cf-isoseg` in AWS you'd just set `next_foundation_in_chain: aws/us-west-2/prod` or something to that effect.

Currently we don't have any "formal rules" to do this, it's just a pattern we've been using for various projects like cf-mgmt, upgrade-tiles, PKS cluster management, etc.; The more I think about it the more I realize there's a pattern here that someone smarter than me could take and run with & formalize... or maybe not. Idk.

## problems encountered

- It creates a lot of github PRs, **but it's actually pretty great for traceability** and better than not having them at all. Maybe we could introduce auto-merging rules at some point?
- Currently this is an approach works best when we assume every value from one state to another should be copied in it's entirely to the next. E.g. you take the entire file contents of `sandbox/product-version` and dump it to `dev/product-version`. But for more nuanced values, like product configuration value yml's, it's a little tricky to **automatically** take a specific key/value combo from the content of `sandbox/vars.yml` and promote it to `dev/vars.yml`.

## ideas someone could run with

- formalizing it a bit somehow? e.g. what if there were a way to define a configuration monorepo's "schema" and "states" and "transition" rules, and some way to leverage those rules via a CLI (and then in use it in ci/cd to create PRs?)
- share whether there are similar projects or whether someone has already written a ton about this?
- use this in (oosely defined) pattern in other contexts, like ... idk, maybe for terraform config? For `helm`-deployment configs? For describing applications deployed on cloud foundry? etc.

## conclusion: "okay cool why did you even write this?"

It's helpful to have these thoughts/use-cases externalized into github issues so we can reference them when we talk about use-cases

Also because I think this is a great usage pattern for configuration monorepos, I think it has a lot of uses in many different contexts, and **I want people smarter than me to see it so they do something similar and help improve the idea, or tell me if there's a better way to do it**

We wrote up semi-related thoughts about how the "monorepo schema" is layered in a post over at cf-mgmt https://github.com/pivotalservices/cf-mgmt/issues/174 

Thanks for your time, feel free to close whenever.

**Comment by jtarchie** [comment]
We are currently going through the process of _figuring out_ what a recommended way of config organization, with promotion in mind, looks like.

Thanks for the feedback, @aegershman.
This is useful. We will be using this in our research as _real world_ example.

**Comment by kcboyle** [comment]
Hi @aegershman! I wanted to follow up with you on this! We've been doing some work with @clhynfield to try to learn about and document some of these patterns. We haven't documented them yet, but found there are 2 structures we'd like to try to document (one uses `config-template`, and the other assumes `staged-config`). In the deeper exploration of config template, we have a tested WIP for the `staged-config` path in the [docs-platform-automation-reference-pipeline-config](https://github.com/pivotal/docs-platform-automation-reference-pipeline-config/tree/single-product) repo. 

It's a little more naive than the implementation suggested/discussed above, and we're still working on how best to document patterns like these in the Platform Automation [docs](https://docs.pivotal.io/platform-automation)

**Comment by kcboyle** [comment]
Hi @aegershman! 

We added some docs to describe the config repo structure mentioned above in our docs. It won't be _officially_ released until Platform Automation 5.0, but is available to peruse on the `develop` [branch now](https://docs.pivotal.io/platform-automation/develop/pipeline-design/configuration-management-strategies.html#multiple-foundations-with-one-repository)
